pycbc/distributions/external.py

```diff
@@ -106,7 +130,108 @@ def from_config(cls, cp, section, variable_args):
+        return cls(params=params, logpdf=logpdf, rvs=rvs, cdfinv=cdfinv)
+
 
+class DistributionFunctionFromFile(External):
+    r"""Evaluating PDF, logPDF, CDF and inverse CDF from the external
+        density function.
 
-__all__ = ['External']
+    Instances of this class can be called like a distribution in the .ini file,
+    when used with `pycbc.distributions.external.External`. Please see the
+    example in the `External` class.
+
+    Parameters
+    ----------
+    parameter : {'file_path', 'column_index'}
+        The path of the external density function's .txt file, and the
+        column index of the density distribution. By default, the first column
+        should be the values of a certain parameter, such as "mass", other
+        columns should be the corresponding density values (as a function of
+        that parameter).
+    \**kwargs :
+        All other keyword args are passed to `scipy.integrate.quad` to control
+        the numerical accuracy of the inverse CDF.
+        If not be provided, will use the default values in `self.__init__`.
+
+    Attributes
+    ----------
+    name : "external_func_fromfile"
+        The name of this distribution.
+
+    Notes
+    -----
+    This class is different from `pycbc.distributions.arbitrary.FromFile`,
+    which needs samples from the hdf file to construct the PDF by using KDE.
+    This class reads in any continuous functions of the parameter.
+    """
+    name = "external_func_fromfile"
+
+    def __init__(self, params=None, file_path=None,
+                 column_index=None, **kwargs):
+        if kwargs.__contains__('cdfinv'):
+            super().__init__(cdfinv=kwargs['cdfinv'])
+        else:
+            super().__init__(cdfinv=not None)
+        self.params = params
+        self.data = np.loadtxt(file_path, unpack=True, skiprows=1)
+        self.column_index = int(column_index)
+        self.epsabs = kwargs.get('epsabs', 1.49e-05)
+        self.epsrel = kwargs.get('epsrel', 1.49e-05)
+        self.x_list = np.linspace(self.data[0][0], self.data[0][-1], 1000)
+        self.interp = {'pdf': callable, 'cdf': callable, 'cdfinv': callable}
+        if not file_path:
+            raise ValueError("Must provide the path to density function file.")
+
+    def _pdf(self, x, **kwargs):
+        """Calculate and interpolate the PDF by using the given density function,
+        then return the corresponding value at the given x."""
+        if self.interp['pdf'] == callable:
+            func_unnorm = scipy_interpolate.interp1d(
+                self.data[0], self.data[self.column_index])
+            norm_const = scipy_integrate.quad(
+                func_unnorm, self.data[0][0], self.data[0][-1],
+                epsabs=self.epsabs, epsrel=self.epsrel, limit=500,
+                **kwargs)[0]
+            self.interp['pdf'] = scipy_interpolate.interp1d(
+                self.data[0], self.data[self.column_index]/norm_const)
```

Comment:
It will shows `TypeError: unsupported operand type(s) for /: 'interp1d' and 'float'`. Or I must make a `for` loop to reweight each point in `func_unnorm`, after that, it still needs interpolation.